src/meta/greedy_load_balancer.cpp

```diff
@@ -784,105 +784,112 @@ void ford_fulkerson::update_flow(int pos,
+    auto potential_moving_size = potential_moving.size();
+    dassert_f(plan_moving <= potential_moving_size,
+              "from({}) to({}) plan({}), can_move({})",
+              from.to_string(),
+              to.to_string(),
+              plan_moving,
+              potential_moving_size);
+
+    while (plan_moving-- > 0) {
+        dsn::gpid selected = select_moving(potential_moving, prev_load, current_load, from, to);
+
+        const partition_configuration &pc = app->partitions[selected.get_partition_index()];
+        auto balancer_result = t_migration_result->emplace(
+            selected, generate_balancer_request(pc, balance_type::move_primary, from, to));
+        dassert_f(balancer_result.second, "gpid({}) already inserted as an action", selected);
+
+        --(*prev_load)[get_disk_tag(from, selected)];
+        ++(*current_load)[get_disk_tag(to, selected)];
+    }
+}
+
+std::list<dsn::gpid> greedy_load_balancer::calc_potential_moving(
+    const std::shared_ptr<app_state> &app, const rpc_address &from, const rpc_address &to)
+{
+    std::list<dsn::gpid> potential_moving;
+    const node_state &ns = t_global_view->nodes->find(from)->second;
+    ns.for_each_primary(app->app_id, [&](const gpid &pid) {
+        const partition_configuration &pc = app->partitions[pid.get_partition_index()];
+        if (is_secondary(pc, to)) {
+            potential_moving.push_back(pid);
         }
+        return true;
+    });
 
-        current = prev[current];
-        std::swap(current_load, prev_load);
-    }
+    return potential_moving;
+}
 
-    for (auto &kv : ml_this_turn) {
-        auto r = t_migration_result->emplace(kv.first, kv.second);
-        dassert(r.second,
-                "gpid(%d.%d) already inserted as an action",
-                kv.first.get_app_id(),
-                kv.first.get_partition_index());
+dsn::gpid greedy_load_balancer::select_moving(std::list<dsn::gpid> &potential_moving,
+                                              disk_load *prev_load,
+                                              disk_load *current_load,
+                                              rpc_address from,
+                                              rpc_address to)
+{
+    std::list<dsn::gpid>::iterator selected = potential_moving.end();
+    int max = std::numeric_limits<int>::min();
+
+    for (auto it = potential_moving.begin(); it != potential_moving.end(); ++it) {
+        int load_difference =
+            (*prev_load)[get_disk_tag(from, *it)] - (*current_load)[get_disk_tag(to, *it)];
+        if (load_difference > max) {
+            max = load_difference;
+            selected = it;
+        }
     }
-    return true;
+
+    dassert_f(selected != potential_moving.end(),
+              "can't find gpid to move from({}) to({})",
+              from.to_string(),
+              to.to_string());
+    auto res = *selected;
+    potential_moving.erase(selected);
+    return res;
```

Comment:
directly return `*selected`?

Reply:
We can't do it like this, because selected will be removed by `potential_moving.erase(selected);`

Reply:
So the remove should't put here, the function only `select`, better add remove logic, better put out

Reply:
Yes, you are right. But in set of cpp, it doesn't support to move *selected from potential_moving. It only support remove iter from it.